import {
  Body,
  Controller,
  Get,
  InternalServerErrorException,
  NotFoundException,
  Patch,
  Post,
  UnprocessableEntityException,
  UseGuards,
} from "@nestjs/common";
import { User } from "@prisma/client";
import { GetUser } from "../auth/decorator";
import { JwtGuard } from "../auth/guard";
import { UsersService } from "./users.service";
import { EditUserDto } from "./dto";
import { ApiTags } from "@nestjs/swagger";
import { validate } from "class-validator";
@ApiTags("users")
@UseGuards(JwtGuard)
@Controller("users")
export class UsersController {
  constructor(private userService: UsersService) {}

  @Get("me")
  async getMe(@GetUser() user: User) {
    const result = await this.userService.leanUserInfo(user);
    console.log(
      "ðŸš€ ~ file: users.controller.ts:47 ~ UsersController ~ getMe ~ result:",
      result
    );
    return result;
  }

  @Patch("me")
  async editUser(@GetUser("user_id") userId: number, @Body() dto: EditUserDto) {
    // ìœ íš¨ì„± ê²€ì‚¬ ìˆ˜í–‰
    const errors = await validate(dto);
    if (errors.length > 0) {
      const errorResponse = {
        message: "Validation failed",
        errors: errors.map((error) => ({
          field: error.property,
          message: Object.values(error.constraints).join(", "),
        })),
      };
      throw new UnprocessableEntityException(errorResponse);
    }
    try {
      const result = await this.userService.editUser(userId, dto);
      console.log(
        "ðŸš€ ~ file: users.controller.ts:47 ~ UsersController ~ editUser ~ result:",
        result
      );

      return result;
    } catch (error) {
      if (error instanceof NotFoundException) {
        throw error;
      }
      throw new InternalServerErrorException();
    }
  }
}
